baseapp/abci.go

```diff
@@ -190,11 +191,30 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 	app.voteInfos = req.LastCommitInfo.GetVotes()
 
 	// call the hooks with the BeginBlock messages
+	wg := new(sync.WaitGroup)
 	for _, streamingListener := range app.abciListeners {
-		if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
-			app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
+		streamingListener := streamingListener // https://go.dev/doc/faq#closures_and_goroutines
+		if streamingListener.HaltAppOnDeliveryError() {
+			// increment the wait group counter
+			wg.Add(1)
+			go func() {
+				// decrement the counter when the go routine completes
+				defer wg.Done()
+				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
+					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
+					app.halt()
+				}
+			}()
+		} else {
+			go func() {
+				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
+					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
+				}
+			}()
```

Comment:
Because this goroutine is spawned and executed asynchronously, it's possible that it's scheduled when `app.deliverState` has one value, and executes when `app.deliverState` has a different value, e.g. representing the next block. This would mean that `ListenBeginBlock` will be called with a `ctx` that doesn't correspond to its sibling `req` and `res` parameters.

Reply:
I've been debugging an `AppHash` mismatch error when I test the plugins on our provenance blockchain. This is most likely the cause.

Our initial thinking here was that we wanted to make calls to multiple listeners asynchronous. I'll test this out with synchronized listeners and report my findings.

Reply:
The primary consequence of making listeners synchronous is that slow listeners can impact node liveness. If HaltAppOnDeliveryError is true, then users are explicitly opting in to this risk, so it's (arguably) OK there. But if that setting is false, then I think it would be a mistake to make delivery synchronous.

The solution here is to produce the complete event when this method is called, and emit that event _value_ to listeners. The event value would need to contain only plain-old-data, i.e. no pointers or references to anything else in the node or app or etc.

Reply:
> I've been debugging an `AppHash` mismatch error when I test the plugins on our provenance blockchain. This is most likely the cause.
> 
> Our initial thinking here was that we wanted to make calls to multiple listeners asynchronous. I'll test this out with synchronized listeners and report my findings.

I tested this and it is not the cause of the `AppHash` error I see on a two node localnet. I disabled all the listeners in abci.go (thinking this was the cause) and I was able to narrow it down to either the [Inject()](https://github.com/provenance-io/cosmos-sdk/blob/1092048c0b7943d0a54b2b91b7f2e1e3cd0b9064/plugin/loader/loader.go#L232) or [Start()](https://github.com/provenance-io/cosmos-sdk/blob/1092048c0b7943d0a54b2b91b7f2e1e3cd0b9064/plugin/loader/loader.go#L250) as the cause. When you turn on the plugin system and enable any plugin it will result in an `AppHash` error. I also used the [iaveiwer](https://github.com/cosmos/iavl/tree/master/cmd/iaviewer) tool but didn't see any differences in the iAVL tree. This leads me to think the state change is happening at runtime before it's persisted.

Something like.
```
panic: Failed to process committed block (11:AF09D0FF0AACA156034874C0A14F3C2A4EB4FF384D22201062AF5E4ADD2FAE4A): wrong Block.Header.AppHash.  Expected 6AC03C53EC307EFA46AA61B006833C9C6A6A37C1B54D8B184D2B1B468D8026AB, got 7003EC3FBBDEADC9D62E27C8DF75B27BDF7EF4B575E5653E8D7921571CB9C64C
```

@i-norden ^^^ perhaps you can take a look at the plugin system and maybe find what is changing the app state. 




> The primary consequence of making listeners synchronous is that slow listeners can impact node liveness. If HaltAppOnDeliveryError is true, then users are explicitly opting in to this risk, so it's (arguably) OK there. But if that setting is false, then I think it would be a mistake to make delivery synchronous.
> 
> The solution here is to produce the complete event when this method is called, and emit that event _value_ to listeners. The event value would need to contain only plain-old-data, i.e. no pointers or references to anything else in the node or app or etc.

I agree ^^^. I was testing to see if asynchronous vs synchronous listeners was the cause for my `AppHash` error above. We'll also need to pass in any addition info required to keep track of what block is being emitted etc. 

so the new listener interface could look like this?
```
type ABCIListener interface {
	ListenBeginBlock(logger log.Logger, blockHeight int64, req []byte, res []byte) error
	ListenEndBlock(logger log.Logger, blockHeight int64, req []byte, res []byte) error
	ListenDeliverTx(logger log.Logger, blockHeight int64, req []byte, res []byte) error
	HaltAppOnDeliveryError() bool
}
```

Reply:
An ABCIListener interface should 

1. Take a `context.Context` as a first parameter, and respect context cancellation
2. Already know about any `log.Logger` it needs, and therefore not take a logger as a parameter
3. Not need to signal HaltAppOnDeliveryError thru a method, but instead either
  a. Halt the app directly via os.Exit or panic or whatever, or
  b. Signal that information via return error from each method, detected by the caller via errors.Is/As